test/apps/iodemo/io_demo.cc

```diff
@@ -832,132 +889,182 @@ class DemoClient : public P2pDemoCommon {
         }
+    }
 
-        // reset number of retries after successful connection
-        _retry = 0;
+    status_t run() {
+        _server_info.resize(opts().servers.size());
+        std::for_each(_server_info.begin(), _server_info.end(),
+                      reset_server_info);
 
         _status = OK;
 
         // TODO reset these values by canceling requests
         _num_sent      = 0;
         _num_completed = 0;
 
-        uint32_t sn          = IoDemoRandom::rand<uint32_t>();
-        double prev_time     = get_time();
-        long total_iter      = 0;
-        long total_prev_iter = 0;
-        std::vector<op_info_t> info;
-
-        for (int i = 0; i < IO_OP_MAX; ++i) {
-            op_info_t op_info = {static_cast<io_op_t>(i), 0, 0};
-            info.push_back(op_info);
-        }
+        uint32_t sn                  = IoDemoRandom::rand<uint32_t>();
+        double prev_time             = get_time();
+        double prev_connect_time     = 0;
+        long total_iter              = 0;
+        long total_prev_iter         = 0;
+        op_info_t op_info[IO_OP_MAX] = {{0,0}};
 
         while ((total_iter < opts().iter_count) && (_status == OK)) {
             VERBOSE_LOG << " <<<< iteration " << total_iter << " >>>>";
 
-            if (!wait_for_responses(opts().window_size - 1)) {
+            wait_for_responses(opts().window_size - 1);
+            if (_status != OK) {
                 break;
             }
 
-            size_t conn_num = IoDemoRandom::rand(size_t(0), conn.size() - 1);
-            io_op_t op      = get_op();
+            /* Always try to reconnect every if all servers are disconnected,
+             * or try every 10 iterations if only some servers are disconnected
+             */
+            if ((_active_servers.size() < _server_info.size()) &&
+                (_active_servers.empty() || ((total_iter % 10) == 0))) {
+                double curr_time = get_time();
+                if (curr_time > (prev_connect_time + opts().client_timeout)) {
+                    prev_connect_time = curr_time;
+                    connect_all();
+                    if (_status != OK) {
+                        break;
+                    }
+                }
+            }
+
+            if (_active_servers.empty()) {
+                LOG << "All servers are disconnected";
+                sleep(opts().client_timeout);
+                continue;
+            }
+
+            /* Pick random connected server */
+            int active_index = IoDemoRandom::rand(size_t(0),
+                                                  _active_servers.size() - 1);
+            size_t server_index = _active_servers[active_index];
+            assert(_server_info[server_index].conn != NULL);
+
+            io_op_t op = get_op();
             size_t size;
             switch (op) {
             case IO_READ:
-                size = do_io_read(conn[conn_num], sn);
+                size = do_io_read(server_index, sn);
```

Comment:
IMO `conn` arg better here but need add a reference to `server_info`

Reply:
i'll change it to passing ref